test: add test for signet miner script

theStack opened the pull request:
This PR adds a very basic test for the signet miner script (contrib/signet/miner). ~~It was based on #24553 (merged by now) which fixes a bug (and was also the motivation to write this test).~~

The test roughly follows the steps from https://en.bitcoin.it/wiki/Signet#Custom_Signet, except that the challenge key-pair is created solely with the test framework. Calibration is also skipped, the difficulty is simply set to the first mainnet target `0x1d00ffff` (see also https://bitcoin.stackexchange.com/a/57186).

theStack committed:
test: add test for signet miner script

laanwj commented:
Concept ACK, thanks for adding tests

laanwj commented on test/functional/tool_signet_miner.py:
```diff
@@ -0,0 +1,72 @@
+        # determine paths to tools
+        base_dir = self.config["environment"]["SRCDIR"]
+        signet_miner_path = os.path.join(base_dir, "contrib", "signet", "miner")
+        # TODO: this is a dirty hack; the path to bitcoin-util should ideally
```
I don't think it's good to leave it like this, if we use `bitcoin-util` in the functional tests we'd at least need to

- Add a `self.options.bitcoinutil` in `BitcoinTestFramework.setup`, analogous to the other two, overridable with `BITCOINUTIL`

- we also need an entry in `config.ini.in` for `BUILD_BITCOIN_UTIL` → an `is_bitcoin_util_compiled()` → and an `skip_if_no_bitcoin_util`

theStack replied:
Thanks for the guide, will add a commit with these preliminaries within the next days.

theStack committed:
test: determine path to `bitcoin-util` in test framework

The path is stored in `self.options.bitcoinutil`, points to
`src/bitcoin-util` by default and can be overrided with the
`BITCOINUTIL` environment variable.

theStack committed:
test: add `is_bitcoin_util_compiled` helper

theStack commented:
Force-pushed with preliminary commits integrating `bitcoin-util` into the test framework (storing its path to `self.options.bitcoinutil`, adding `is_bitcoin_util_compiled`/`skip_if_no_bitcoin_util` helpers depending on the `test/config.ini` entry), [as suggested by @laanwj](https://github.com/bitcoin/bitcoin/pull/24559#discussion_r842715566).

laanwj commented:
Code review ACK f8abc66ca16eb0c3a4fa2388b11b9bef3c8a297a

laanwj commented on test/functional/tool_signet_miner.py:
```diff
@@ -0,0 +1,63 @@
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_cli()
+        self.skip_if_no_wallet()
+        self.skip_if_no_bdb()
```
Is there a specific reason to require bdb here? I don't think it even specifically makes a bdb wallet?

theStack replied:
The problem was the signing in the descriptor wallet rather than the import of the private key. Now it should work with both wallet types: https://github.com/bitcoin/bitcoin/pull/24559#issuecomment-1098546332

0xB10C commented:
Concert ACK!

theStack commented:
Force-pushed with a simpler signet challenge (p2wpkh instead of 1-of-1 multisig that the descriptor wallet couldn't sign), the test can now be run with both legacy and descriptor wallet. Also removed the "works only with legacy wallet" part from the PR description accordingly.

laanwj commented:
re-ACK 038d2a607f93a31be76f3a8210a5385996012783
Thanks for making the test work for the non-legacy wallet too.
